constraints/miden-vm/stack.air

```diff
@@ -0,0 +1,957 @@
+    enf s[i]' = s[i] for i in 5..16
+
+ev movup5(main: [s[16]]):
+    enf s[0]' = s[5]
+    enf s[i + 1]' = s[i] for i in 0..5
+    enf s[i]' = s[i] for i in 6..16
+
+ev movdn5(main: [s[16]]):
+    enf s[5]' = s[0]
+    enf s[i]' = s[i + 1] for i in 0..5
+    enf s[i]' = s[i] for i in 6..16
+
+ev movup6(main: [s[16]]):
+    enf s[0]' = s[6]
+    enf s[i + 1]' = s[i] for i in 0..6
+    enf s[i]' = s[i] for i in 7..16
+
+ev movdn6(main: [s[16]]):
+    enf s[6]' = s[0]
+    enf s[i]' = s[i + 1] for i in 0..6
+    enf s[i]' = s[i] for i in 7..16
+
+ev movup7(main: [s[16]]):
+    enf s[0]' = s[7] = 0
+    enf s[i + 1]' = s[i] for i in 0..7
+    enf s[i]' = s[i] for i in 8..16
+
+ev movdn7(main: [s[16]]):
+    enf s[7]' = s[0]
+    enf s[i]' = s[i + 1] for i in 0..7
+    enf s[i]' = s[i] for i in 8..16
+
+ev swapw(main: [s[16]]):
+    enf s[i]' = s[i + 4] for i in 0..4
+    enf s[i + 4]' = s[i] for i in 0..4
+    enf s[i]' = s[i] for i in 8..16
+
+ev ext2mul(main: [s[16]]):
+    enf s[0]' = s[0]
+    enf s[1]' = s[1]
+    enf s[2]' = (s[0] + s[1]) * (s[2] + s[3]) - s[0] * s[2]
+    enf s[3]' = s[1] * s[3] - 2 * s[0] * s[2]
+    enf s[i]' = s[i] for i in 4..16
+
+ev movup8(main: [s[16]]):
+    enf s[0]' = s[8]
+    enf s[i + 1]' = s[i] for i in 0..8
+    enf s[i]' = s[i] for i in 9..16
+
+ev movdn8(main: [s[16]]):
+    enf s[8]' = s[0]
+    enf s[i]' = s[i + 1] for i in 0..8
+    enf s[i]' = s[i] for i in 9..16
+
+ev swapw2(main: [s[16]]):
+    enf s[i]' = s[i + 8] for i in 0..4
+    enf s[i + 8]' = s[i] for i in 0..4
+    enf s[i]' = s[i] for i in 4..8
+    enf s[i]' = s[i] for i in 12..16
+
+# Should we enforce that elements in position grater than 15 do not change? 
+ev swapw3(main: [s[16]]):
+    enf s[i]' = s[i + 12] for i in 0..4
+    enf s[i + 12]' = s[i] for i in 0..4
+    enf s[i]' = s[i] for i in 4..12
+
+# Should we enforce that elements in position grater than 15 do not change? 
+ev swapdw(main: [s[16]]):
+    enf s[i]' = s[i + 8] for i in 0..8
+    enf s[i + 8]' = s[i] for i in 0..8
+
+
+# Left stack shift operationds
+
+ev assert(main: [s[16]]):
+    enf s[0] = 1
+    enf s[i]' = s[i + 1] for i in 0..15
+
+ev eq(main: [s[16]], aux: [h0]):
+    enf s[0]' * (s[0] - s[1]) = 0
+    enf s[0]' = 1 - (s[0] - s[1]) * h0
+    enf s[i]' = s[i + 1] for i in 1..15
+
+ev add(main: [s[16]]):
+    enf s[0]' = s[0] + s[1]
+    enf s[i]' = s[i + 1] for i in 1..15
+
+ev mul(main: [s[16]]):
+    enf s[0]' = s[0] * s[1]
+    enf s[i]' = s[i + 1] for i in 1..15
+
+ev and(main: [s[16]]):
+    enf is_binary([s[i]]) = 0 for i in 0..2
+    enf s[0]' = s[0] * s[1]
+    enf s[i]' = s[i + 1] for i in 1..15
+
+ev or(main: [s[16]]):
+    enf is_binary([s[i]]) = 0 for i in 0..2
+    enf s[0]' = s[1] + s[0] - s[1] * s[0]
+    enf s[i]' = s[i + 1] for i in 1..15
+
+ev u32and(main: [s[16]], aux: [b_chip]):
+    let op_u32and = 2
+    enf b_chip' * ($alpha[0] + $alpha[1] * op_u32and + $alpha[2] * s[0] + $alpha[3] * s[1] + $alpha[4] * s[0]') = b_chip
```

Comment:
We'll probably need to put bus constraints into a separate module as they need to include both left and right sides of the bus, and here we are computing only the left side.

I'm thinking that we'd probably end up having each bus constraints being a separate module.